tests/test_slips_utils.py

```diff
@@ -1,10 +1,410 @@
+    "timestamp, expected_result",
+    [
+        ("1680788096.789", "1680788096"),
+        ("1680788096", "1680788096"),
+        ("1680788096.", "1680788096"),
+    ],
+)
+def test_remove_milliseconds_decimals(timestamp, expected_result):
+    utils = ModuleFactory().create_utils_obj()
+    result = utils.remove_milliseconds_decimals(timestamp)
+    assert (
+        result == expected_result
+    ), f"Expected {expected_result}, but got {result}"
+
+
+@pytest.mark.parametrize(
+    "start_time, end_time, return_type, expected_result",
+    [
+        (1609459200, 1609545600, "days", 1),
+        (1609459200, 1609545600, "hours", 24),
+        (1609545600, 1609459200, "seconds", lambda result: result < 0),
+        (1609459200.0, 1609459200.1, "seconds", 0.1),
+    ],
+)
+def test_get_time_diff(start_time, end_time, return_type, expected_result):
+    utils = ModuleFactory().create_utils_obj()
+    result = utils.get_time_diff(start_time, end_time, return_type)
+    if callable(expected_result):
+        assert expected_result(
+            result
+        ), f"Expected the result to satisfy the condition, but got {result}"
+    else:
+        assert (
+            result == expected_result
+        ), f"Expected {expected_result}, but got {result}"
+
+
+@pytest.mark.parametrize(
+    "seconds, expected_timedelta",
+    [
+        (3600, datetime.timedelta(seconds=3600)),
+        (86400, datetime.timedelta(days=1)),
+        (-3600, datetime.timedelta(seconds=-3600)),
+        (31536000, datetime.timedelta(days=365)),
+    ],
+)
+def test_to_delta(seconds, expected_timedelta):
+    utils = ModuleFactory().create_utils_obj()
+    result = utils.to_delta(seconds)
+    assert (
+        result == expected_timedelta
+    ), f"Expected {expected_timedelta}, but got {result}"
+
+
+@pytest.mark.parametrize(
+    "side_effect, expected_result",
+    [
+        (None, lambda utils: isinstance(utils.get_own_IPs(), list)),
+        (
+            requests.exceptions.ConnectionError,
+            lambda utils: "127.0.0.1" in utils.get_own_IPs()
+            or not utils.get_own_IPs(),
+        ),
+        (
+            requests.exceptions.RequestException,
+            lambda utils: "127.0.0.1" in utils.get_own_IPs()
+            or not utils.get_own_IPs(),
+        ),
+    ],
+)
+def test_get_own_IPs(side_effect, expected_result):
+    utils = ModuleFactory().create_utils_obj()
+    if side_effect:
+        with patch("requests.get", side_effect=side_effect):
+            assert expected_result(
+                utils
+            ), "Should handle the situation gracefully"
+    else:
+        assert expected_result(utils), "Should return a list of IPs"
+
+
+@pytest.mark.parametrize("port", [80, 65535, 0])
+def test_is_port_in_use(port):
+    utils = ModuleFactory().create_utils_obj()
+    assert isinstance(utils.is_port_in_use(port), bool)
+
+    if port != 0:
+        assert (
+            utils.is_port_in_use(port) is False
```

Comment:
not every non-zero port should be reported as not used. i suggest we test it using another way, not sure how to test it though, let's search!

Reply:
Okay